File: frontend/src/types/transaction.ts
export interface Transaction {
  id: string;
  amount: number;
  description: string;
  transactionDate: string;
  type: TransactionType;
  status: TransactionStatus;
  accountId: string;
  accountName: string;
  categoryId?: string;
  categoryName?: string;
  reference?: string;
  notes?: string;
  isRecurring: boolean;
  tags: TransactionTag[];
  createdAt: string;
}

export enum TransactionType {
  Income = 0,
  Expense = 1,
  Transfer = 2,
}

export enum TransactionStatus {
  Pending = 0,
  Completed = 1,
  Failed = 2,
  Cancelled = 3,
}

export interface TransactionTag {
  id: string;
  name: string;
  color: string;
}

export interface CreateTransactionRequest {
  amount: number;
  description: string;
  transactionDate: string;
  type: TransactionType;
  accountId: string;
  categoryId?: string;
  reference?: string;
  notes?: string;
  isRecurring: boolean;
  recurrencePattern?: string;
  tagNames: string[];
}

export interface UpdateTransactionRequest {
  amount?: number;
  description?: string;
  transactionDate?: string;
  type?: TransactionType;
  status?: TransactionStatus;
  accountId?: string;
  categoryId?: string;
  reference?: string;
  notes?: string;
  isRecurring?: boolean;
  recurrencePattern?: string;
  tagNames?: string[];
}

export interface TransactionFilter {
  fromDate?: string;
  toDate?: string;
  type?: TransactionType;
  status?: TransactionStatus;
  accountId?: string;
  categoryId?: string;
  minAmount?: number;
  maxAmount?: number;
  search?: string;
  page: number;
  pageSize: number;
  sortBy: string;
  sortOrder: string;
}

export interface TransactionSummary {
  totalIncome: number;
  totalExpense: number;
  netAmount: number;
  transactionCount: number;
  fromDate: string;
  toDate: string;
}

export interface PagedResult<T> {
  items: T[];
  totalCount: number;
  page: number;
  pageSize: number;
  totalPages: number;
}

File: backend/src/SmartFinance.Domain/Entities/Transaction.cs
using SmartFinance.Domain.Common;
using SmartFinance.Domain.Enums;

namespace SmartFinance.Domain.Entities;

public class Transaction : BaseAuditableEntity
{
    public decimal Amount { get; set; }
    public string Description { get; set; } = string.Empty;
    public DateTime TransactionDate { get; set; } = DateTime.UtcNow;
    public TransactionType Type { get; set; }
    public TransactionStatus Status { get; set; } = TransactionStatus.Pending;
    public Guid AccountId { get; set; }
    public Guid UserId { get; set; }
    public Guid? CategoryId { get; set; }
    public string? ExternalId { get; set; }
    public string? Reference { get; set; }
    public decimal? ExchangeRate { get; set; }
    public string? Notes { get; set; }
    public bool IsRecurring { get; set; }
    public string? RecurrencePattern { get; set; }

    public virtual Account Account { get; set; } = null!;
    public virtual User User { get; set; } = null!;
    public virtual Category? Category { get; set; }
    public virtual ICollection<TransactionTag> Tags { get; set; } = new List<TransactionTag>();
}
